ambry-tools/src/main/java/com.github.ambry/store/DumpIndexTool.java

```diff
@@ -42,481 +49,358 @@
- * 5. Dump N random blobs for a replica
+ * 1. Dump index.
+ * 2. Dump an index segment.
+ * 3. Verify an index for sanity (no duplicate records).
  */
 public class DumpIndexTool {
-  private final ClusterMap clusterMap;
-  // set to true if only error logging is required
-  private final boolean silent;
+
+  public class Info {
+    private final EnumSet<BlobState> states;
+    private final boolean isInRecentIndexSegment;
+
+    public Info(EnumSet<BlobState> states, boolean isInRecentIndexSegment) {
+      this.states = states;
+      this.isInRecentIndexSegment = isInRecentIndexSegment;
+    }
+
+    public EnumSet<BlobState> getStates() {
+      return states;
+    }
+
+    public boolean isInRecentIndexSegment() {
+      return isInRecentIndexSegment;
+    }
+  }
+
+  public class IndexProcessingResults {
+    private final Map<StoreKey, Info> keyToState;
+    private final long processedCount;
+    private final long putCount;
+    private final long deleteCount;
+    private final Set<StoreKey> duplicatePuts;
+    private final Set<StoreKey> putAfterDeletes;
+    private final Set<StoreKey> duplicateDeletes;
+    private final long activeCount;
+
+    public IndexProcessingResults(Map<StoreKey, Info> keyToState, long processedCount, long putCount, long deleteCount,
+        Set<StoreKey> duplicatePuts, Set<StoreKey> putAfterDeletes, Set<StoreKey> duplicateDeletes) {
+      this.keyToState = keyToState;
+      this.processedCount = processedCount;
+      this.putCount = putCount;
+      this.deleteCount = deleteCount;
+      this.duplicatePuts = duplicatePuts;
+      this.putAfterDeletes = putAfterDeletes;
+      this.duplicateDeletes = duplicateDeletes;
+      activeCount = keyToState.values().stream().filter(value -> value.getStates().contains(BlobState.Valid)).count();
+    }
+
+    public Map<StoreKey, Info> getKeyToState() {
+      return keyToState;
+    }
+
+    public long getProcessedCount() {
+      return processedCount;
+    }
+
+    public long getActiveCount() {
+      return activeCount;
+    }
+
+    public long getPutCount() {
+      return putCount;
+    }
+
+    public long getDeleteCount() {
+      return deleteCount;
+    }
+
+    public Set<StoreKey> getDuplicatePuts() {
+      return duplicatePuts;
+    }
+
+    public Set<StoreKey> getPutAfterDeletes() {
+      return putAfterDeletes;
+    }
+
+    public Set<StoreKey> getDuplicateDeletes() {
+      return duplicateDeletes;
+    }
+
+    public boolean isIndexSane() {
+      return duplicatePuts.size() == 0 && putAfterDeletes.size() == 0 && duplicateDeletes.size() == 0;
+    }
+
+    @Override
+    public String toString() {
+      return "Processed: " + processedCount + ", Active: " + activeCount + ", Put: " + putCount + ", Delete: "
+          + deleteCount + ", Duplicate Puts: " + duplicatePuts + ", Put After Delete: " + putAfterDeletes
+          + ", Duplicate Delete: " + duplicateDeletes;
+    }
+  }
+
+  public enum BlobState {
+    Valid, Deleted, Expired;
```

Comment:
not in this patch. Can you file an issue to add Corrupted state here

Reply:
I'm not sure how `Corrupted` makes sense in the `DumpIndexTool`. What is a corrupted index entry? The `IndexSegment` wouldn't even deserialize in that case, right?

Reply:
Oops, sorry. I was having log dump in mind when I made this comment.